# References for transcriptomics

1. Wang, Zhong, Mark Gerstein, and Michael Snyder. "RNA-Seq: a revolutionary 
   tool for transcriptomics." Nature reviews genetics 10.1 (2009): 57-63.
2. Stark, Rory, Marta Grzelak, and James Hadfield. "RNA sequencing: the teenage 
   years." Nature Reviews Genetics 20.11 (2019): 631-656.
3. Conesa, Ana, et al. "A survey of best practices for RNA-seq data analysis." 
   Genome biology 17.1 (2016): 13.
4. Su, Zhenqiang, et al. "A comprehensive assessment of RNA-seq accuracy, 
   reproducibility and information content by the Sequencing Quality Control 
   Consortium." Nature biotechnology 32.9 (2014): 903-914.
5. Pratapa, Aditya, et al. "Benchmarking algorithms for gene regulatory network 
   inference from single-cell transcriptomic data." Nature Methods 17.2 (2020): 
   147-154.
6. Zhao, Shanrong, Zhan Ye, and Robert Stanton. "Misuse of RPKM or TPM 
   normalization when comparing across samples and sequencing protocols." RNA 
   (2020): rna-074922.
7. Roberts, Adam, et al. "Improving RNA-Seq expression estimates by correcting 
   for fragment bias." Genome biology 12.3 (2011): 1-14.
8. Love, Michael I., John B. Hogenesch, and Rafael A. Irizarry. "Modeling of 
   RNA-seq fragment sequence bias reduces systematic errors in transcript 
   abundance estimation." Nature biotechnology 34.12 (2016): 1287.
9. Love, Michael I., Wolfgang Huber, and Simon Anders. "Moderated estimation of 
   fold change and dispersion for RNA-seq data with DESeq2." Genome biology 
   15.12 (2014): 550.
10. Schurch, Nicholas J., et al. "How many biological replicates are needed in 
    an RNA-seq experiment and which differential expression tool should you use?." 
    Rna 22.6 (2016): 839-851.
11. Liu, Yuwen, Jie Zhou, and Kevin P. White. "RNA-seq differential expression 
    studies: more sequence or more replication?." Bioinformatics 30.3 (2014): 
    301-304.